### PR TITLE
async template rendering

### PR DESCRIPTION
**Pull request by scien**
## PROOF OF CONCEPT: DO NOT MERGE

opening for discussion
## 

remaining problems with this implementation
- can't do nested rendering
  - with the ability for templates to be asynchronous, we lose the ability to simply store the previous buffer, render an inner template, grab the html and restore the buffer. 
  - nested rendering can be accomplished, but we would need to be smarter about storing state
- don't have tag names in the stack trace when an error is thrown
  - this implementation sets up a linear todo list, and then executes them in order. it's also possible that tags get added to the queue sequentially instead of children first in the case a given tag is asynchronous. the queue knows what kind of tag it is, but i'm unfamiliar with how to setup named functions in coffee-script to get something like `at Teacup.(anonymous function) [as p]` - anybody have experience with this?
  - i've been a thinking parsing with a tree may be a much more flexible implementation. each tag appends it's opening, contents, and closing tags as children. and then contents generates it's own children. and everything propagates up the tree when when rendering is complete. i essentially do that inline here, but a tree implementation may allow us to render multiple asynchronous tags at once. we would still have to modify teacup to not use a simple `htmlOut` buffer to write to, as rendering child templates and rendering multiple async tags at once all need their own local buffers.
## 

testing: this code passes 81 tests, and fails 4. the 4 failures are...
- render nested in a template returns the nested template without clobbering the parent result
- render nested in a template is aliased as cede:
- renderable decorator supports composition with renderable and non-renderable helpers:
- stack trace should contain tag names
## 

my preferred syntax for async in teacup

```
render ( ->
  div (done) ->
    setTimeout ( ->
      text 'hello world'
      done()
    ), 1
), (html) ->
```

any `tag` can use a single function argument has a callback function to let the caller know the async tag is complete.

something more real life...

```
render ( ->
  div '.articles', (done) ->
    db.articles.find {category: 'coffee-script'}, (err, articles) ->
      if err
        text err
      else
        for article in articles
          div '.article', ->
            article.name
      done()
), (html) ->
```

note: the `</div>` closing tag will render when `done()` is called, so `done` must be run at the end of the asynchronous function after rendering your contents.
## 

running this code: to get an idea for how it works i left in some code for extra logging, go to `Queue::constructor` and set `@debug = true`. this will show you functions as they get added to the queue and run.

here are two examples

```
    html = render ( ->
      div ->
        'hello world'
    )
```

it will go through the template and queue up each `raw`, `text`, or `contents`.

```
[
  "raw: <div>",
  "contents: function () {\n          return 'hello world';\n        }",
  "raw: </div>"
]
```

it then processes the queue, in this case. `raw` adds `<div>` to `htmlOut`. then `contents` runs and inserts a `text "hello world"` at the beginning of the queue, because we need to render inner content before rendering the closing tag. the biggest trick to getting this to work was to reset the position i insert into the queue to 0 every time `contents` is run. this allows all inner content to be processed before the closing tag, and before any following tags.

```
[
  "text: hello world",
  "raw: </div>"
]
```

it then renders `"hello world"` and `</div>`

this makes asynchronous possible, because we don't run the next function in the queue until the current one is done.  so let's assume `contents` above was an asynchronous function.

```
[
  "contents: function (done) {\n          return setTimeout((function() {\n            text('hello world');\n            return done();\n          }), 1);\n        }",
  "raw: </div>"
]
```

we just had to have a callback given to the function to know when it completed.
## 

note: templates can use an argument that is not a function and it will know it's not a asynchronous tag callback parameter

```
  it 'allows template arguments', ->
    template = (user) ->
      div ->
        user.name

    html = render template, {name: 'bryant'}
    expect(html).to.equal '<div>bryant</div>'
```
## 

it will also handle renderable components correctly

```
  it 'lets components provide give contents to their children', ->
    modal = component (selector, attrs, renderContents) ->
      closeButton = ->
        button 'Close'

      div '.modal', ->
        renderContents(closeButton)

    template = ->
      modal (closeButton) ->
        text 'close me: '
        closeButton()

    expect(render template).to.equal '<div class="modal">close me: <button>Close</button></div>'
```

although the contents accept a single function as input, teacup checks to see if a function was passed in as input. it only assumes it's an asynchronous tag if there is no function input and there is one function argument for the contents to be rendered.

note: i think that logic needs some work and more verbose tests for complete code coverage.

### CODE REVIEW

File: src/teacup.coffee [hunk 1]
```diff
@@ -43,6 +45,44 @@ merge_elements = (args...) ->
...
+
+  constructor: (options) ->
+    @debug = false
+    @items = []
+    @names = []
+    @position = 0
+    @running = false
+
+  push: (name, fn) ->
+    @names.splice @position, 0, name
```
Comment (scien) [inline src/teacup.coffee]: names was simply for debugging, but I think maybe we could find a way to use it to get function names back in stack traces

File: src/teacup.coffee [hunk 2]
```diff
@@ -110,11 +147,19 @@ class Teacup
...
-    else
-      @text contents
+    @queue.push "contents: #{contents}", (done) =>
+      @queue.position = 0
+      if typeof contents is 'function'
+        if contents.length is 0 or contents.length is rest.length
+          result = contents.apply @, rest
+          @text result if typeof result is 'string'
+          done()
+        else if contents.length is 1
```
Comment (scien) [inline src/teacup.coffee]: this should probably be something more like `if contents.length is rest.length + 1`

**Comment by scien** [comment]
see https://github.com/scien/teacup/blob/master/test/async.coffee for a list of new tests

**Comment by scien** [comment]
I'd also like to note that when doing client-side rendering, using the synchronous approach is better. instead of using 
```
render ( ->
  div '.articles', (done) ->
    db.articles.find {category: 'coffee-script'}, (err, articles) ->
      if err
        text err
      else
        for article in articles
          div '.article', ->
            article.name
      done()
), (html) ->
```

you should definitely do something more like this
```
$el.html = teacup.render ( ->
  div '.articles', ->
    'loading...'
)
db.articles.find {category: 'coffee-script'}, (err, articles) ->
  $el.find('.articles').html ( ->
    if err
      err
    else
      for article in articles
        div '.article', ->
          article.name
```
so that you can render part of your templates without waiting for data.

The asynchronous rendering that this PR allows will be super convenient for pre-generating html though or when blocking on the entire template isn't an issue, and it allows for a really simple, powerful templating language without needing to jump between template and controller logic

**Comment by scien** [comment]
@hurrymaplelad

**Comment by hurrymaplelad** [comment]
@scien cool proof of concept!  I'm struggling to come up with examples where I'd encourage folks to push async code into a template.   Seems like for small templates, like your client side example, resolving async work before rendering is straightforward.  For larger templates, encapsulating the data retrieval is valuable.

If we want more conventions around async work I lean towards a layer on top of teacup.  For example:

```coffee
class ComplicatedView
  render: (done) ->
    buffer = ''
    @fetchArticle (err, article) =>
      buffer += @renderArticle article
      @fetchComments article, (err, comments) =>
        buffer += @renderComments comments
        done buffer

  renderArticle: renderable (article) ->
    h1 article.title
    p article.body

  renderComments: renderable (comments) ->
    ul ->
      comments.forEach (comment) ->
        li comment.text

  fetchArticles: (done) ->
    request.get 'example.com/articles', done

  fetchComments: (article, done) ->
    request.get "example.com/articles/#{article.id}/comments", done

```

If you're set on baking async control flow into templates, it'll probably be easiest to iterate on in a fork.  The single global buffer is definitely the big issue to tackle.  Changing the way it works will likely change the interface for helpers built on nested templates.  I'd explore passing a non-global buffer between nested templates bound to `this`, and making the buffer a an array containing strings and Promises, concatenating when all promises have resolved.

**Pull request closed without merging**